core/src/main/scala/org/apache/spark/storage/DiskBlockObjectWriter.scala

```diff
@@ -206,18 +209,22 @@ private[spark] class DiskBlockObjectWriter(
         streamOpen = false
         closeResources()
       }
+   } catch {
+      case e: Exception =>
+        logError("Uncaught exception while closing file " + file, e)
+    }
 
-      val truncateStream = new FileOutputStream(file, true)
-      try {
-        truncateStream.getChannel.truncate(committedPosition)
-        file
-      } finally {
-        truncateStream.close()
-      }
+    var truncateStream: FileOutputStream = null
+    try {
+      truncateStream = new FileOutputStream(file, true)
+      truncateStream.getChannel.truncate(committedPosition)
+      file
     } catch {
       case e: Exception =>
         logError("Uncaught exception while reverting partial writes to file " + file, e)
         file
+    } finally {
+      truncateStream.close()
```

Comment:
Another option is to use `Utils.tryWithSafeFinally` and do the if block with closeResources in the block with truncate in the finally block